packages/react/src/utils/useCustomMenuItems.tsx

```diff
@@ -0,0 +1,209 @@
+import { logErrorInDevMode } from '@clerk/shared';
+import type { CustomMenuItem } from '@clerk/types';
+import type { ReactElement } from 'react';
+import React from 'react';
+
+import { MenuAction, MenuItems, MenuLink, UserProfileLink, UserProfilePage } from '../components/uiComponents';
+import {
+  customMenuItemsIgnoredComponent,
+  userButtonIgnoredComponent,
+  userButtonMenuItemLinkWrongProps,
+  userButtonMenuItemsActionWrongsProps,
+} from '../errors/messages';
+import type { UserButtonActionProps, UserButtonLinkProps } from '../types';
+import type { UseCustomElementPortalParams, UseCustomElementPortalReturn } from './useCustomElementPortal';
+import { useCustomElementPortal } from './useCustomElementPortal';
+
+const isThatComponent = (v: any, component: React.ReactNode): v is React.ReactNode => {
+  return !!v && React.isValidElement(v) && (v as React.ReactElement)?.type === component;
+};
+
+export const useUserButtonCustomMenuItems = (children: React.ReactNode | React.ReactNode[]) => {
+  const reorderItemsLabels = ['manageAccount', 'signOut'];
+  return useCustomMenuItems({
+    children,
+    reorderItemsLabels,
+    MenuItemsComponent: MenuItems,
+    MenuActionComponent: MenuAction,
+    MenuLinkComponent: MenuLink,
+    UserProfileLinkComponent: UserProfileLink,
+    UserProfilePageComponent: UserProfilePage,
+    componentName: 'UserButton',
+  });
+};
+
+type UseCustomMenuItemsParams = {
+  children: React.ReactNode | React.ReactNode[];
+  MenuItemsComponent?: any;
+  MenuActionComponent?: any;
+  MenuLinkComponent?: any;
+  UserProfileLinkComponent?: any;
+  UserProfilePageComponent?: any;
+  componentName: string;
+  reorderItemsLabels: string[];
+};
+
+type CustomMenuItemType = UserButtonActionProps | UserButtonLinkProps;
+
+const useCustomMenuItems = ({
+  children,
+  MenuItemsComponent,
+  MenuActionComponent,
+  MenuLinkComponent,
+  UserProfileLinkComponent,
+  UserProfilePageComponent,
+  reorderItemsLabels,
+  componentName,
+}: UseCustomMenuItemsParams) => {
+  const validChildren: CustomMenuItemType[] = [];
+  const customMenuItems: CustomMenuItem[] = [];
+  const customMenuItemsPortals: React.ComponentType[] = [];
+
+  React.Children.forEach(children, child => {
+    if (
+      !isThatComponent(child, MenuItemsComponent) &&
+      !isThatComponent(child, UserProfileLinkComponent) &&
+      !isThatComponent(child, UserProfilePageComponent)
+    ) {
+      if (child) {
+        logErrorInDevMode(userButtonIgnoredComponent(componentName));
+      }
+      return;
+    }
+
+    // Ignore UserProfileLinkComponent and UserProfilePageComponent
+    if (isThatComponent(child, UserProfileLinkComponent) || isThatComponent(child, UserProfilePageComponent)) {
+      return;
+    }
+
+    // Menu items children
+    const { props } = child as ReactElement;
+
+    React.Children.forEach(props.children, child => {
```

Comment:
❓Is this how we are getting access to the children of `<User.MenuItems/>` ?

Reply:
I don't like the nested traversal, but i can't think of something better

Reply:
> ❓Is this how we are getting access to the children of `<User.MenuItems/>` ?

Yes @panteliselef, we are doing this in order to get access to the children of `<User.MenuItems/>`